File: expressionAdd.py
#Recursive solution
#Complexities: exponential
result = list()
targ = 0
def addOperators(num,target):

    if len(num) == 0 or num is None:
        return result
    targ = target
    helper(num,"",0,0,0)
    return result

def helper(num,path,calc,tail,index):
    if(index==len(num)):
        if calc == targ:
            result.add(path)
        return
    for i in range(index,len(num)):
        curr = num[index:i+1]
        if i == 0:
            helper(num,path+curr,curr,curr,i+1)
        else:
            helper(num,path+"+"+curr,calc+curr,curr,i+1)
            helper(num,path+"-"+curr,calc+curr,int(curr),i+1)
            helper(num,path+"*"+curr,calc-tail+tail*curr,tail*curr,i+1)
st = "123"
targ = 6
addOperators(st,targ)

File: expressionAdd.cpp
class Solution {

    vector<string> result;
public:
    vector<string> addOperators(string num, int target) {
        
        helper(num,target,0,"",0,0);
        return result;
    }

    public:

    void helper(string num, int target, int pivot, string path, long calc, long tail)
    {
        if(pivot == num.length() && calc == target)
        {
            result.push_back(path);
        }

        for(int i = pivot; i<num.length(); i++)
        {
            if(i != pivot && num.at(pivot) == '0')
            continue;

            long curr = long(num.substr(pivot, i+1));
            
                if(pivot == 0)
                {
                    helper(num,target, i+1 , path + curr, curr, curr);
                }
                else
                {
                    //+
                    helper(num,target, i+1, path+"+"+curr, calc + curr, curr);

                    //-
                    helper(num,target, i+1, path+"-"+curr, calc - curr, -curr);

                    //*
                    helper(num,target, i+1, path+"*"+curr, calc - tail +(tail * curr), tail*curr);

                }
            
        }
    }
};